core/string/ustring.cpp

```diff
@@ -1639,13 +1639,43 @@ Vector<int> String::split_ints_mk(const Vector<String> &p_splitters, bool p_allo
 }
 
 String String::join(const Vector<String> &parts) const {
+	if (parts.is_empty()) {
+		return String();
+	} else if (parts.size() == 1) {
+		return parts[0];
+	}
+
+	const int this_length = length();
+
+	int new_size = (parts.size() - 1) * this_length;
+	for (const String &part : parts) {
+		new_size += part.length();
+	}
+	new_size += 1;
+
 	String ret;
-	for (int i = 0; i < parts.size(); ++i) {
-		if (i > 0) {
-			ret += *this;
+	ret.resize(new_size);
+	char32_t *ret_ptrw = ret.ptrw();
+	const char32_t *this_ptr = ptr();
+
+	bool first = true;
+	for (const String &part : parts) {
+		if (first) {
```

Comment:
Does compiler optimize this out? If not, taking it out of the loop might be an option for an ounce more (have to profile though). :slightly_smiling_face:

Reply:
I assume it'll unroll it

Reply:
Ah what is it they say about assumption.. :wink: 

It's only a minor thing anyway but ime optimizers don't always make the decisions you expect, as they don't have knowledge of the data, they can't assume you are feeding a list of zero items, 1 or 2, or 10000. Good practice to look at the assembly or look in godbolt (or just profile and see what is going on).

Having said that it should be an improvement so far anyway so happy to approve.

Reply:
Tested a simple case with gcc and it unrolls a loop like this one, by jumping past the check and then looping back around, like this for a trivial case:
```assembly
        cmpq    %rbx, %rsi
        jne     .L6
        jmp     .L1
        .p2align 4,,10
        .p2align 3
.L8:
        movl    (%rbx), %ecx
        call    _Z11call_me_tooi
.L6:
        movl    (%rbx), %ecx
        addq    $4, %rbx
        call    _Z7call_mei
        cmpq    %rbx, %rsi
        jne     .L8
.L1:
        addq    $40, %rsp
        popq    %rbx
        popq    %rsi
        ret
```

This is from this code:
```cpp
void do_this(const std::vector<int> &p_arg) {
    bool first = true;
    for (const int &i : p_arg) {
        if (first) {
            first = false;
        } else {
            call_me_too(i);
        }
        call_me(i);
    }
}
```
So it will jump to L6 which is the `call_me` line, and then next time it loops it'll go to `call_me_too`